### ISSUE TITLE
Get data directly from the database

### ISSUE THREAD

**Issue by maurolepore** [2DegreesInvesting/ds-incubator#39]
## Who is the audience?

Analysts, data managers, software developers at 2DII and beyond.

## Why is this important?

You may want to get data as close as possible to its ultimate source. That way you can control how and when you process raw data to get analysis data.

## What should be covered?

* Where does this fit in the data-science workflow?
* How this approach compares to what we already do, for example, what are the advantages and dissadvantages?

## Suggested speakers or contributors

@cjyetman volunteered to lead this meetup. I will give a 5' introduction.

## Resources

* ...

### 2h before

- [ ] Open meetup issue, paste this template and edit as necessary
- [ ] Link resources
- [ ] Develop content for the ds-incubator website
- [ ] Announce on #coding
- [ ] Add meetup date, title and link to the invitation on 2dii's calendar
- [ ] Set alarm for +15' into the meeting
- [ ] Set alarm for +25' into the meeting

### 10' before

- [ ] Remind meeting and share link on #coding
- [ ] Login at <https://zoom.us/> and start recurring ds-incubator meeting`***`
    * `***` Backup (Google) https://meet.google.com/azf-xifn-scd
- [ ] Under _Participants > More_ uncheck "Put participants in waiting room ..."

### Start

- [ ] 00': Welcome
- [ ] 02': Announcements
- [ ] 05': Record
- [ ] 15': Q&A + discuss "must-have" or "nice-to-have"
- [ ] 25': End
- [ ] 30': Upload video
- [ ] 35': Link video under resources
- [ ] 35': Label as "must-have" or "nice-to-have"

**Comment by cjyetman** [comment]
@tposey28 noted in Slack that technically the host (URL) in the connection should be `twodii-database.westeurope.cloudapp.azure.com`. The host in the script I used in the presentation is using the staging DB, not the production DB.

**Comment by cjyetman** [comment]
one thought on the last comment that @Clare2D made which we didn't have time to properly address...

In our current process, if a change is made to a specific set of quarterly data in the DB, a copy of the previous state plus a copy of the new state are both saved in Dropbox with a timestamp identifier.

I presume that a similar, or parallel process could be done in the database, creating a new schema with a version number or timestamp, so that each version of the quarterly data was available in separate schemas... and probably a quarterly schema that would always be the latest/most-up-to-date version. Something like...
`_2019q4` = latest version of _2019q4 data
`_2019q4_20200102` = version of _2019q4 data at 2020.01.02
`_2019q4_20200213` = version of _2019q4 data at 2020.02.13

and then you could access them like...
`datastore <- connect_datastore(data_quarter = '_2019q4')`
`datastore <- connect_datastore(data_quarter = '_2019q4_20200102')`
`datastore <- connect_datastore(data_quarter = '_2019q4_20200213')`

**Comment by maurolepore** [comment]
Thanks @cjyetman for following up. 

I still know nothing of that process but your suggestion seems good and elegant. That seems to be a common approach to pulling the `latest` or `specific` versions of software and data -- for example it's how we get Docker images, GitHub releases.

It would be good to discuss with the database team, then analysts, and come up with a solution what works for everyone. I'll follow up.

**Comment by maurolepore** [comment]
@Clare2D, @2diiKlaus, @koopmand and other experienced analysts, do you see any other issue we need to consider?

**Comment by tposey28** [comment]
CJ, that's a good idea. The database would get fairly large, but I think
it's the best solution we have until Alex and I build out a data warehouse
(which solves this issue in a far my elegant way).

The one problem I see is that there is a schema where we store all the
non-time dependent data, things like the list of companies and securities.
All the quarterly schemas refer to this data in order to maintain
consistent IDs. In your solution there is a caveat that the data today
versus next week could still be slightly different for a schema. Nothing
major should change, since the IDs should stay the same, but a company may
have an altered name or a security an altered issuer in the current
structure, since I store those in that schema. I could try to change that
by changing where I store those data points, but it's probably okay as long
as it's noted. Will look into it.

**Comment by cjyetman** [comment]
Thanks @tposey28 

I definitely recognize that adding new schemas for every set of changes adds some overhead, and would require some strategic planning. Just want to disambiguate between capabilities and procedural changes. What is currently done with changes to a quarterly set of data is a procedural thing, not a specific capability of Dropbox versus database access. That procedure could be modified/adapted so that the same or similar capability would exist with database access also.

**Comment by tposey28** [comment]
I completely agree; we can change our workflow to create solutions to the
historical data question, we're not tethered to Dropbox because we need
file versioning.

Mauro and I just had a talk about this which changed my stance a bit. I can
make a schema for every change, but what I consider to be fundamental
changes and what the analyst considers to be fundamental changes are
different, and I can't have a schema for every small change I make. Imagine
a database that changes daily (or hourly). There should be database
snapshots, certainly, but they don't capture every change, and it is up to
the analysts to capture the data they need when they need it (along with
all previous versions they used).

So I believe the best solution is something in-between: Database admins
(me/Alex) create versions reaching back to a certain point (say the last 3
versions) as well as a future version (dev work) since recent changes may
cause mistakes and should be easily compared to the previous version. Then,
analysts capture the data for a project, timestamping and storing for
future reference, then review the changes that have taken place (through
our change logs, but hopefully through a project specific QA process as
well). This takes the responsibility of predicting which data change is
important off of the data admins, whose roles should primarily be spent on
keeping the data current.

Basically, I'm going to move forward with your idea (I think it holds a lot
of value), but hopefully we can construct a workflow that removes the need.

**Comment by 2diiKlaus** [comment]
The capability of going back to previous versions of the data by adding a version number as suggested by CJ is an absolute must at the current state. 
Basically "analysts" need to be able to go back in time in terms of data & code. How we achieved this is open to discuss and looking at CJs suggestion that sounds like a good plan. 

Alternatively, the analysts would need to create a copy of the data for each project, which kind of works against our "improve the data structure & flow" work as well as can limit the capabilities to assess further things. Which might also be fine - but needs to be communicated quite well to project leads so that they do not promise things we cannot do.

That said (as mauro asked for input from analysts): I support the idea from CJ (short term solution) and are happy to discuss a long term solution separately

**Comment by tposey28** [comment]
Klaus, I definitely see the importance. However, consider all other data
providers. A up and running version for all previous states of data just
doesn't exist in a standard database workflow. Backups for changes is
common, but those are not actively running. Currently, if the server has to
hold more than 10 versions we'll run into memory issues. We can upgrade the
machine at that point, but then that's a change to our budget.

I'm certain there is a workflow where it is not counter intuitive for
analysts to store data for a specific project (not an export like what I
do, just the couple of files touched by that project). This issue isn't a
2dii exclusive one, someone has solved it and I believe we just haven't
implemented it yet.

Until then though I'll look into to maintaining old versions going forward.
I will not be able to do this retroactively,
 or at least not for all versions. If there is a historical state that is
absolutely critical I can find a backup.

**Comment by cjyetman** [comment]
@tposey28 would it make sense to collect on-going incremental changes in one staging schema (or separate DB), and release tagged versions on a less frequent basis to the production DB? Definitely agree it doesn't, make sense to make a whole new schema every time a super minor change is made, but maybe if every 2 weeks or whatever a new tagged version of a given quarter's data was pushed to the production DB with all the changes that were made since the last... something like that? I would imagine if that was the understood process, analysts would prefer to use one of the tagged versions in the production DB, unless they really wanted some micro change that happened in the last few days, and in that case it would be clear in their mind that they were using an intermediate version that would be difficult to replicate if necessary?

**Comment by maurolepore** [comment]
Thanks! I'll discuss by technology:

* Git is great for small, text based files but not limited to them. It can track large files including binaries. The diff will be uninformative but you can time-travel binaries or text files the same way -- with `git checkout <commit, tag (version), branch, etc.>` (or using your favorite GUI). [StackOverflow suggests](https://stackoverflow.com/questions/3230721/max-file-size-and-max-repository-size-in-git) points to specialized tools for extreemly large files (never tried [this](http://caca.zoy.org/wiki/git-bigfiles)).

* GitHub can host individual files up to 2GB. How you move files there vary by size, by [the pins package hides the different ways](https://pins.rstudio.com/articles/boards-github.html#large-files) so you need not worry. (It's as usual for files size 25MB or less, and it's as attachments to a GitHub-release for larger files.). Here is [an example where pins is used to create a data pipeline](https://pins.rstudio.com/articles/use-cases/create-data-pipelines.html) that we could use as a guide to build ours.

* Azure has no limit in file size, so we could use it limitless (except for cost) as a board to pin our gigantic datasets, and any number of versions -- which would have a local cache.
    * [Azure boards](https://pins.rstudio.com/articles/boards-azure.html)
    * [Versioning boards](https://pins.rstudio.com/articles/advanced-versions.html)

Here I name technologies to make things concrete and to pass what I recently learned about pins. But I am open to learning other frameworks and support people adopt whatever system works for them.

I hope to digest this in an end-to-end example sometime soon.

**Comment by maurolepore** [comment]
> I'm certain there is a workflow where it is not counter intuitive for analysts to store data for a specific project .... This issue isn't a 2dii exclusive one, someone has solved it and I believe we just haven't implemented it yet.
-- @tposey28 

I suspect the same. I'll see if we can get free consultation with an RStudio representative. They might be interested in selling us something but if it's a useful product then why not listen?
